fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/mv/AbstractMaterializedViewRule.java

```diff
@@ -568,6 +563,93 @@ protected List<Expression> rewriteExpression(List<? extends Expression> sourceEx
+     * type#1 -> type#9
+     * payload#4 -> payload#10
+     * query variants is payload['issue']['number']#20
+     * then we can add payload['issue']['number']#20 -> element_at(element_at(payload#10, 'issue'), 'number')
+     * to targetToTargetReplacementMappingQueryBased
+     * */
+    private void extendMappingByVariant(Set<SlotReference> queryVariants,
+            Map<Expression, Expression> targetToTargetReplacementMappingQueryBased) {
+        if (queryVariants.isEmpty()) {
+            return;
+        }
+        Map<List<String>, Expression> viewNameToExprMap = new HashMap<>();
+        for (Map.Entry<Expression, Expression> targetExpressionEntry :
+                targetToTargetReplacementMappingQueryBased.entrySet()) {
+            if (targetExpressionEntry.getKey() instanceof SlotReference
+                    && ((SlotReference) targetExpressionEntry.getKey()).getDataType() instanceof VariantType) {
+                List<String> nameIdentifier = new ArrayList<>();
+                nameIdentifier.add(((SlotReference) targetExpressionEntry.getKey()).getName());
+                nameIdentifier.addAll(((SlotReference) targetExpressionEntry.getKey()).getSubPath());
+                viewNameToExprMap.put(nameIdentifier, targetExpressionEntry.getValue());
+            }
+        }
+        if (viewNameToExprMap.isEmpty()) {
+            return;
+        }
+        Map<List<String>, SlotReference> queryNameAndExpressionMap = new HashMap<>();
+        for (SlotReference slotReference : queryVariants) {
+            List<String> nameIdentifier = new ArrayList<>();
+            nameIdentifier.add(slotReference.getName());
+            nameIdentifier.addAll(slotReference.getSubPath());
+            queryNameAndExpressionMap.put(nameIdentifier, slotReference);
+        }
+        for (Map.Entry<List<String>, ? extends Expression> queryNameEntry : queryNameAndExpressionMap.entrySet()) {
+            Expression minExpr = null;
+            List<String> minCompensateName = null;
+            for (Map.Entry<List<String>, Expression> entry : viewNameToExprMap.entrySet()) {
+                if (!containsAllWithOrder(queryNameEntry.getKey(), entry.getKey())) {
+                    continue;
+                }
+                List<String> removedQueryName = new ArrayList<>(queryNameEntry.getKey());
+                removedQueryName.removeAll(entry.getKey());
+                if (minCompensateName == null) {
+                    minCompensateName = removedQueryName;
+                    minExpr = entry.getValue();
+                }
+                if (removedQueryName.size() < minCompensateName.size()) {
+                    minCompensateName = removedQueryName;
+                    minExpr = entry.getValue();
+                }
+            }
+            if (minExpr != null) {
+                targetToTargetReplacementMappingQueryBased.put(queryNameEntry.getValue(),
+                        constructElementAt(minExpr, minCompensateName));
```

Comment:
after rewrite, these new element_at will be push into scan node too?

Reply:
yeah, this will push into scan node again by RBO